File: Mathematics/Combination Calculation/leetcode/62. Unique Paths/GPT/Unique-Paths.py
# # ✅ 回答

# ## 1. 問題の分析

# ### 競技プログラミング視点

# * **高速化優先**

#   * 単純な DFS/BFS は指数時間 O(2^(m+n)) で非現実的。
#   * DP（動的計画法）: O(m·n)、空間 O(m·n) だが、制約 (m,n ≤ 100) では十分。
#   * 数学的アプローチ（二項係数 C(m+n-2, m-1)）を使えば O(min(m,n)) で計算可能。
# * **メモリ効率**

#   * DP は 2D 配列が必要だが、O(min(m,n)) に圧縮可能。
#   * 数学的アプローチなら定数メモリ O(1)。

# ### 業務開発視点

# * 可読性・保守性 → DP 実装が直感的で理解しやすい。
# * 数学的アプローチは数式に依存するため、背景知識が必要。
# * エラーハンドリング: m, n の範囲チェック、型チェックを入れる。

# ### Python特有の考慮点

# * **CPythonの特性**

#   * 組み込み `math.comb` は C 実装で超高速。
#   * ループ処理よりも `math.comb` を使うほうが格段に速い。
# * **GILの影響** → 単一スレッドなので無関係。
# * **内蔵データ構造** → DP 実装時は `list` を使えば十分。

# ---

# ## 2. アルゴリズムアプローチ比較

# | アプローチ           | 時間計算量                              | 空間計算量       | Python実装コスト | 可読性 | 標準ライブラリ活用 |
# | --------------- | ---------------------------------- | ----------- | ----------- | --- | --------- |
# | DP (2D 配列)      | O(m·n)                             | O(m·n)      | 低           | 高   | list      |
# | DP (1D 圧縮)      | O(m·n)                             | O(min(m,n)) | 中           | 中   | list      |
# | 数学的アプローチ (二項係数) | O(min(m,n)) or O(1 with math.comb) | O(1)        | 低           | 中   | math.comb |

# ---

# ## 3. 選択したアルゴリズムと理由

# * **競技プログラミング**: `math.comb` を用いた **二項係数アプローチ**。

#   * O(1) 時間、O(1) 空間で最速。
# * **業務開発**: DP アプローチも併せて実装。

#   * 可読性が高く、チーム開発で理解しやすい。
# * **Python特有の最適化**

#   * `math.comb` を利用（Python 3.8+）。
#   * DP 実装は内包表記とリスト更新で高速化。

# ---

# ## 4. コード実装

# ```python
from typing import List
import math

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        """
        動的計画法による実装（業務開発向け）
        
        Args:
            m: 行数 (1 <= m <= 100)
            n: 列数 (1 <= n <= 100)
            
        Returns:
            ロボットが右下まで到達するユニークパス数
        
        Raises:
            ValueError: m, n が範囲外の場合
            TypeError: m, n が int でない場合
            
        Time Complexity: O(m*n)
        Space Complexity: O(n) （1次元圧縮DP）
        """
        if not isinstance(m, int) or not isinstance(n, int):  # type: ignore[unnecessary-isinstance]
            raise TypeError("m and n must be integers")
        if not (1 <= m <= 100 and 1 <= n <= 100):
            raise ValueError("m and n must be within [1, 100]")

        # DP: 1次元配列で管理
        dp: List[int] = [1] * n
        for _ in range(1, m):
            for j in range(1, n):
                dp[j] += dp[j-1]
        return dp[-1]

    def uniquePaths_optimized(self, m: int, n: int) -> int:
        """
        数学的アプローチ（二項係数）による最適化版（競技プログラミング向け）
        
        Args:
            m: 行数
            n: 列数
        
        Returns:
            ユニークパス数
        
        Time Complexity: O(1) （math.comb利用）
        Space Complexity: O(1)
        """
        # C(m+n-2, m-1)
        return math.comb(m+n-2, m-1)
# ```

# ---

# ## 5. 制約条件への対応

# * 外部ライブラリは未使用 → `math` のみ。
# * メモリ使用量は `O(n)` 以下。
# * Python 3.11+ → `math.comb` は利用可能。

# ---

# ✅ **まとめ**

# * 業務開発では **DP（1次元圧縮）** が理解しやすく安全。
# * 競技プログラミングでは **数学的アプローチ** が最速。

# ---
# Pylance が `isinstance(m, int)` / `isinstance(n, int)` に対して **UnnecessaryIsInstance** を出しているんですね。

# これは **型ヒントで引数が `int` と既に宣言されている** 場合に、`isinstance` チェックが不要だと静的解析器が判断しているためです。

# ---

# ## ✅ 解決方法

# ### 1. 競技プログラミング向け（最小限）

# * 型は入力制約で保証されている（`int`）とみなし、`isinstance` を削除してOK。

# ```python
# class Solution:
#     def uniquePaths(self, m: int, n: int) -> int:
#         if not (1 <= m <= 100 and 1 <= n <= 100):
#             raise ValueError("m and n must be within [1, 100]")
        
#         dp: list[int] = [1] * n
#         for _ in range(1, m):
#             for j in range(1, n):
#                 dp[j] += dp[j-1]
#         return dp[-1]

#     def uniquePaths_optimized(self, m: int, n: int) -> int:
#         import math
#         return math.comb(m + n - 2, m - 1)
# ```

# ---

# ### 2. 業務開発向け（堅牢性を残す場合）

# * **外部入力（例: JSON, CLI 引数, DB など）から呼び出される可能性**があるなら `isinstance` を残して良い。
# * ただし、Pylance の警告を避けたいなら **コメントで明示的に抑制**可能。

# ```python
# if not isinstance(m, int) or not isinstance(n, int):  # type: ignore[unnecessary-isinstance]
#     raise TypeError("m and n must be integers")
# ```

# ---

# ### 3. 実務でのベストプラクティス

# * **内部で完結する競技プログラム → 削除**
# * **外部入力が混ざる業務コード → 残す（ただし type: ignore）**

File: Mathematics/Combination Calculation/leetcode/62. Unique Paths/Claude/Unique-Paths.js
// ## 1. 問題の分析

// ### 競技プログラミング視点での分析
// - **実行速度最優先**: 数学的解法（組み合わせ計算）が最速 - O(min(m,n))
// - **メモリ最小化**: 1次元配列によるDP - O(min(m,n))の空間計算量

// ### 業務開発視点での分析
// - **保守性重視**: 2次元DPテーブルによる直感的な実装
// - **エラーハンドリング**: 入力値の範囲チェック、オーバーフロー検証

// ### JavaScript特有の考慮点
// - **V8最適化**: 単型配列（数値のみ）の使用、インデックスアクセス優先
// - **GC対策**: 大きな2次元配列の生成回避、プリミティブ値の活用
// - **精度問題**: Number.MAX_SAFE_INTEGER（2^53-1）以下での計算保証

// ## 2. アルゴリズムアプローチ比較

// |アプローチ|時間計算量|空間計算量|JS実装コスト|可読性|備考|
// |---------|---------|---------|-----------|-----|---|
// |数学的解法(組み合わせ)|O(min(m,n))|O(1)|低|中|最高速、オーバーフロー注意|
// |1次元DP|O(m×n)|O(min(m,n))|低|高|メモリ効率良好|
// |2次元DP|O(m×n)|O(m×n)|中|最高|直感的、メモリ使用量大|
// |再帰+メモ化|O(m×n)|O(m×n)|高|中|スタックオーバーフロー危険|

// ## 3. 選択したアルゴリズムと理由

// ### 選択したアプローチ: 数学的解法（組み合わせ計算）

// ### 理由:
// - **計算量的優位性**: O(min(m,n))の時間計算量で最も高速
// - **JavaScript環境での効率**: プリミティブ値のみ使用でGC負荷最小
// - **制約条件適合**: m,n ≤ 100なら精度問題なし

// ### JavaScript特有の最適化ポイント:
// - BigIntを使わずNumber型で十分（制約範囲内）
// - ループ変数の型統一でV8最適化促進
// - 条件分岐最小化でインライン化促進## JavaScript固有の最適化観点

// ### V8エンジン最適化のポイント
// 1. **インライン化促進**:
//    - 小さな関数設計でJITコンパイラによる最適化
//    - 単純なforループ使用（forEachより高速）

// 2. **Hidden Classの安定化**:
//    - 配列は数値のみで型統一
//    - オブジェクト生成を最小化

// 3. **数値計算の最適化**:
//    - Number型での整数演算（BigInt回避）
//    - 型変換の最小化

// ### ガベージコレクション対策
// 1. **メモリ使用量最小化**:
//    - 数学的解法では追加メモリ不要（O(1)空間）
//    - DP解法でも1次元配列のみ使用

// 2. **オブジェクト生成回避**:
//    - プリミティブ値のみ使用
//    - 配列サイズ事前確保で再割り当て防止

// ### 実装の特徴
// - **主要解法**: 組み合わせ数学を活用したO(min(m,n))解法
// - **副次解法**: メモリ効率の良い1次元DP
// - **エラー処理**: 型チェック、範囲チェック完備
// - **ベンチマーク**: パフォーマンス比較機能付き

// この実装は制約条件（m,n ≤ 100）下で最適なパフォーマンスを実現し、JavaScript環境での実行効率を最大化しています。

/**
 * ロボットがグリッドの左上から右下へ移動する経路数を計算
 * 数学的解法（組み合わせ計算）を使用してO(min(m,n))で解決
 *
 * @param {number} m - グリッドの行数
 * @param {number} n - グリッドの列数
 * @returns {number} 一意な経路の総数
 * @throws {TypeError} 引数が数値でない場合
 * @throws {RangeError} 引数が制約範囲外の場合
 * @complexity Time: O(min(m,n)), Space: O(1)
 */
function uniquePaths(m, n) {
    // 入力検証
    if (typeof m !== 'number' || typeof n !== 'number') {
        throw new TypeError('Arguments must be numbers');
    }

    if (!Number.isInteger(m) || !Number.isInteger(n)) {
        throw new TypeError('Arguments must be integers');
    }

    if (m < 1 || n < 1 || m > 100 || n > 100) {
        throw new RangeError('Arguments must be in range [1, 100]');
    }

    // 数学的解法: C(m+n-2, min(m-1, n-1))を計算
    // 総移動回数: m+n-2 (右にn-1回、下にm-1回)
    // そのうち右移動(またはdown移動)を選ぶ組み合わせ数

    const totalMoves = m + n - 2;
    const rightMoves = n - 1;
    const downMoves = m - 1;

    // 計算効率のため、小さい方を選択
    const k = Math.min(rightMoves, downMoves);

    // C(totalMoves, k) = totalMoves! / (k! * (totalMoves-k)!)
    // オーバーフローを避けるため逐次計算
    let result = 1;

    // V8最適化: 単純なforループを使用
    for (let i = 0; i < k; i++) {
        // result = result * (totalMoves - i) / (i + 1)
        // 整数除算を保証するため、先に乗算してから除算
        result = (result * (totalMoves - i)) / (i + 1);
    }

    return Math.round(result); // 浮動小数点誤差対策
}

/**
 * 動的プログラミング解法（1次元配列使用）
 * メモリ効率を重視した実装
 *
 * @param {number} m - グリッドの行数
 * @param {number} n - グリッドの列数
 * @returns {number} 一意な経路の総数
 * @throws {TypeError} 引数が数値でない場合
 * @throws {RangeError} 引数が制約範囲外の場合
 * @complexity Time: O(m*n), Space: O(min(m,n))
 */
function uniquePathsDP(m, n) {
    // 入力検証（共通関数化可能だが、パフォーマンス重視で直書き）
    if (typeof m !== 'number' || typeof n !== 'number') {
        throw new TypeError('Arguments must be numbers');
    }

    if (!Number.isInteger(m) || !Number.isInteger(n)) {
        throw new TypeError('Arguments must be integers');
    }

    if (m < 1 || n < 1 || m > 100 || n > 100) {
        throw new RangeError('Arguments must be in range [1, 100]');
    }

    // メモリ効率のため、小さい方の次元で配列を作成
    const cols = Math.min(m, n);
    const rows = Math.max(m, n);

    // V8最適化: 配列サイズを事前確保し、単型配列を維持
    const dp = new Array(cols);

    // 初期化: 最初の行は全て1
    for (let j = 0; j < cols; j++) {
        dp[j] = 1;
    }

    // DPテーブル更新
    for (let i = 1; i < rows; i++) {
        for (let j = 1; j < cols; j++) {
            // dp[j] = dp[j] + dp[j-1]
            // 上からの経路 + 左からの経路
            dp[j] += dp[j - 1];
        }
    }

    return dp[cols - 1];
}

/**
 * パフォーマンステスト用のベンチマーク関数
 */
function benchmark() {
    const testCases = [
        [3, 7], // Example 1
        [3, 2], // Example 2
        [23, 12], // 中規模
        [100, 100], // 最大サイズ
    ];

    console.log('Performance Benchmark:');
    console.log('=====================');

    testCases.forEach(([m, n]) => {
        console.log(`\nTest case: m=${m}, n=${n}`);

        // 数学的解法のベンチマーク
        const start1 = performance.now();
        const result1 = uniquePaths(m, n);
        const end1 = performance.now();

        // DP解法のベンチマーク
        const start2 = performance.now();
        const result2 = uniquePathsDP(m, n);
        const end2 = performance.now();

        console.log(`Mathematical: ${result1} (${(end1 - start1).toFixed(4)}ms)`);
        console.log(`DP Solution:  ${result2} (${(end2 - start2).toFixed(4)}ms)`);
        console.log(`Results match: ${result1 === result2}`);
    });
}

// 使用例とテスト
console.log('=== Robot Unique Paths Solution ===\n');

// Example 1
console.log(`uniquePaths(3, 7) = ${uniquePaths(3, 7)}`); // 28

// Example 2
console.log(`uniquePaths(3, 2) = ${uniquePaths(3, 2)}`); // 3

// エラーハンドリングテスト
try {
    uniquePaths(0, 5);
} catch (error) {
    console.log(`Error handling: ${error.message}`);
}

try {
    uniquePaths(101, 5);
} catch (error) {
    console.log(`Error handling: ${error.message}`);
}

// パフォーマンステスト実行
// benchmark(); // 必要に応じてコメントアウト解除

// モジュールエクスポート（Node.js環境用）
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        uniquePaths,
        uniquePathsDP,
        benchmark,
    };
}
